### PR TITLE
Rewriter uses bounds information to add constraints for _Ptr types

### PR DESCRIPTION
**Pull request by awruef**
The rewriter can now check for bounds safe interfaces and update the constraints based on what it finds. Right now, it adds constraints that are not understood by the constraint solving algorithm, and making the constraint solving algorithm understand them is next.

### CODE REVIEW

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 1]
```diff
@@ -51,9 +62,9 @@ PointerVariableConstraint::PointerVariableConstraint(const QualType &QT, uint32_
 				std::pair<uint32_t, Qualification>(K, ConstQualification));
 
 		K++;
-        std::string TyName = tyToStr(Ty);
-        // TODO: Github issue #61: improve handling of types for
-        // variable arguments.
+    std::string TyName = tyToStr(Ty);
+    // TODO: Github issue #61: improve handling of types for
+    // // variable arguments.
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: Looks like a spurious change.

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 2]
```diff
@@ -42,7 +42,18 @@ PointerVariableConstraint::PointerVariableConstraint(const QualType &QT, uint32_
...
+   
+    if (Ty->isCheckedPointerType()) {
+      if (Ty->isCheckedPointerPtrType()) {
+        // Constrain V so that it can't be either wild or an array.
+        CS.addConstraint(CS.createNot(CS.createEq(V, CS.getArr()))); 
+        CS.addConstraint(CS.createNot(CS.createEq(V, CS.getWild())));
+        ConstrainedVars.insert(K);
+      } else if (Ty->isCheckedPointerArrayType()) {
+        llvm_unreachable("unsupported!");
+      }
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: This is only handling bounds-safe interfaces of _Ptr type.  You can also have bounds-safe interfaces of _Array_ptr type.   It's fine to break this into two separate commits but the description of the commit should be modified.

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 3]
```diff
@@ -201,8 +218,18 @@ FunctionVariableConstraint::FunctionVariableConstraint(const Type *Ty,
     const FunctionProtoType *FT = Ty->getAs<FunctionProtoType>();
     assert(FT != nullptr); 
     returnType = FT->getReturnType();
-    for (unsigned i = 0; i < FT->getNumParams(); i++) 
-      paramTypes.push_back(FT->getParamType(i));
+    for (unsigned i = 0; i < FT->getNumParams(); i++) {
+      QualType QT;
+
+      if(const BoundsExpr *BE = FT->getParamBounds(i))
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: It would be good to add a comment explaining what is going on here.

File: tools/checked-c-convert/ProgramInfo.h [hunk 4]
```diff
@@ -66,7 +67,7 @@ class ConstraintVariable {
   virtual void dump() const = 0;
 
   // Constrain everything 'within' this ConstraintVariable to be equal to C.
-  virtual void constrainTo(Constraints &CS, ConstAtom *C) = 0;
+  virtual void constrainTo(Constraints &CS, ConstAtom *C, bool checkSkip=false) = 0;
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.h]: Could you add a comment describing what checkSkip does?

File: tools/checked-c-convert/ProgramInfo.h [hunk 5]
```diff
@@ -54,6 +54,7 @@ class ConstraintVariable {
   ConstraintVariableKind Kind;
 protected:
   std::string BaseType;
+  std::set<uint32_t> ConstrainedVars;
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.h]: Please add a comment about this set.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 1 above)
Comment (awruef) [inline tools/checked-c-convert/ProgramInfo.cpp]: This is actually changing whitespace to match with the rest of the file.

File: tools/checked-c-convert/ProgramInfo.h (hunk 5 above)
Comment (awruef) [inline tools/checked-c-convert/ProgramInfo.h]: Done.

File: tools/checked-c-convert/ProgramInfo.h (hunk 4 above)
Comment (awruef) [inline tools/checked-c-convert/ProgramInfo.h]: Done.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 2 above)
Comment (awruef) [inline tools/checked-c-convert/ProgramInfo.cpp]: I changed the title of the PR, I'll be sure to mention this in the squash and merge comments.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 3 above)
Comment (awruef) [inline tools/checked-c-convert/ProgramInfo.cpp]: Done.

**Comment by awruef** [comment]
I think I addressed your comments in this most recent commit.
